### Ordinal Number Encoding

In [43]:
import datetime

In [44]:
today_date = datetime.datetime.today()

In [45]:
today_date

datetime.datetime(2024, 2, 25, 13, 35, 29, 783049)

In [46]:
today_date-datetime.timedelta(1)

datetime.datetime(2024, 2, 24, 13, 35, 29, 783049)

In [47]:
today_date-datetime.timedelta(2)

datetime.datetime(2024, 2, 23, 13, 35, 29, 783049)

In [50]:
#### List Comprehension
[today_date-datetime.timedelta(x) for x in range(0,15)]

[datetime.datetime(2024, 2, 25, 13, 35, 29, 783049),
 datetime.datetime(2024, 2, 24, 13, 35, 29, 783049),
 datetime.datetime(2024, 2, 23, 13, 35, 29, 783049),
 datetime.datetime(2024, 2, 22, 13, 35, 29, 783049),
 datetime.datetime(2024, 2, 21, 13, 35, 29, 783049),
 datetime.datetime(2024, 2, 20, 13, 35, 29, 783049),
 datetime.datetime(2024, 2, 19, 13, 35, 29, 783049),
 datetime.datetime(2024, 2, 18, 13, 35, 29, 783049),
 datetime.datetime(2024, 2, 17, 13, 35, 29, 783049),
 datetime.datetime(2024, 2, 16, 13, 35, 29, 783049),
 datetime.datetime(2024, 2, 15, 13, 35, 29, 783049),
 datetime.datetime(2024, 2, 14, 13, 35, 29, 783049),
 datetime.datetime(2024, 2, 13, 13, 35, 29, 783049),
 datetime.datetime(2024, 2, 12, 13, 35, 29, 783049),
 datetime.datetime(2024, 2, 11, 13, 35, 29, 783049)]

In [49]:
days=[today_date-datetime.timedelta(x) for x in range(0,15)]

In [51]:
import pandas as pd
data=pd.DataFrame(days)
data.columns=["Day"]

In [52]:
data.head()

Day
0 2024-02-25 13:35:29.783049
1 2024-02-24 13:35:29.783049
2 2024-02-23 13:35:29.783049
3 2024-02-22 13:35:29.783049
4 2024-02-21 13:35:29.783049

In [57]:
data['weekday']=data['Day'].dt.day_name()
data.head()

Day    weekday
0 2024-02-25 13:35:29.783049     Sunday
1 2024-02-24 13:35:29.783049   Saturday
2 2024-02-23 13:35:29.783049     Friday
3 2024-02-22 13:35:29.783049   Thursday
4 2024-02-21 13:35:29.783049  Wednesday

In [58]:
dictionary={'Monday':1,'Tuesday':2,'Wednesday':3,'Thursday':4,'Friday':5,'Saturday':6,'Sunday':7
    
    
}

In [59]:
dictionary

{'Monday': 1,
 'Tuesday': 2,
 'Wednesday': 3,
 'Thursday': 4,
 'Friday': 5,
 'Saturday': 6,
 'Sunday': 7}

In [65]:
data['weekday'].map(dictionary)

0     7
1     6
2     5
3     4
4     3
5     2
6     1
7     7
8     6
9     5
10    4
11    3
12    2
13    1
14    7
Name: weekday, dtype: int64

In [63]:
data['weekday_ordinal']=data['weekday'].map(dictionary)

In [64]:
data

Day    weekday  weekday_ordinal
0  2024-02-25 13:35:29.783049     Sunday                7
1  2024-02-24 13:35:29.783049   Saturday                6
2  2024-02-23 13:35:29.783049     Friday                5
3  2024-02-22 13:35:29.783049   Thursday                4
4  2024-02-21 13:35:29.783049  Wednesday                3
5  2024-02-20 13:35:29.783049    Tuesday                2
6  2024-02-19 13:35:29.783049     Monday                1
7  2024-02-18 13:35:29.783049     Sunday                7
8  2024-02-17 13:35:29.783049   Saturday                6
9  2024-02-16 13:35:29.783049     Friday                5
10 2024-02-15 13:35:29.783049   Thursday                4
11 2024-02-14 13:35:29.783049  Wednesday                3
12 2024-02-13 13:35:29.783049    Tuesday                2
13 2024-02-12 13:35:29.783049     Monday                1
14 2024-02-11 13:35:29.783049     Sunday                7

In [39]:
data['Day'].dt.day #(month,year,hour )

0     25
1     24
2     23
3     22
4     21
5     20
6     19
7     18
8     17
9     16
10    15
11    14
12    13
13    12
14    11
Name: Day, dtype: int32

In [21]:
### chatgpt
from datetime import datetime

# Get the current date
current_date = datetime.now()

# Get the weekday as an integer (Monday is 0, Sunday is 6)
weekday_number = current_date.weekday()

# Get the weekday name
weekday_name = current_date.strftime('%A')

print("Today is", weekday_name)


Today is Sunday


### Count Or frequency encoding

In [80]:
train_set = pd.read_csv('http://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data' , header = None,index_col=None)
train_set.head()                                                         

0                  1       2           3   4                    5   \
0  39          State-gov   77516   Bachelors  13        Never-married   
1  50   Self-emp-not-inc   83311   Bachelors  13   Married-civ-spouse   
2  38            Private  215646     HS-grad   9             Divorced   
3  53            Private  234721        11th   7   Married-civ-spouse   
4  28            Private  338409   Bachelors  13   Married-civ-spouse   

                   6               7       8        9     10  11  12  \
0        Adm-clerical   Not-in-family   White     Male  2174   0  40   
1     Exec-managerial         Husband   White     Male     0   0  13   
2   Handlers-cleaners   Not-in-family   White     Male     0   0  40   
3   Handlers-cleaners         Husband   Black     Male     0   0  40   
4      Prof-specialty            Wife   Black   Female     0   0  40   

               13      14  
0   United-States   <=50K  
1   United-States   <=50K  
2   United-States   <=50K  
3   United-States   <=50K  
4            Cuba   <=50K

In [81]:
columns =[1,3,5,6,7,8,9,13] # categorical columns

In [82]:
train_set[columns]

1            3                    5   \
0              State-gov    Bachelors        Never-married   
1       Self-emp-not-inc    Bachelors   Married-civ-spouse   
2                Private      HS-grad             Divorced   
3                Private         11th   Married-civ-spouse   
4                Private    Bachelors   Married-civ-spouse   
...                  ...          ...                  ...   
32556            Private   Assoc-acdm   Married-civ-spouse   
32557            Private      HS-grad   Married-civ-spouse   
32558            Private      HS-grad              Widowed   
32559            Private      HS-grad        Never-married   
32560       Self-emp-inc      HS-grad   Married-civ-spouse   

                       6               7       8        9               13  
0            Adm-clerical   Not-in-family   White     Male   United-States  
1         Exec-managerial         Husband   White     Male   United-States  
2       Handlers-cleaners   Not-in-family   White     Male   United-States  
3       Handlers-cleaners         Husband   Black     Male   United-States  
4          Prof-specialty            Wife   Black   Female            Cuba  
...                   ...             ...     ...      ...             ...  
32556        Tech-support            Wife   White   Female   United-States  
32557   Machine-op-inspct         Husband   White     Male   United-States  
32558        Adm-clerical       Unmarried   White   Female   United-States  
32559        Adm-clerical       Own-child   White     Male   United-States  
32560     Exec-managerial            Wife   White   Female   United-States  

[32561 rows x 8 columns]

In [105]:
train_set=train_set[columns]

KeyError: "None of [Index([1, 3, 5, 6, 7, 8, 9, 13], dtype='int32')] are in the [columns]"

In [108]:
### Creating our own columns
train_set.columns=['Employment','Degree', 'Status','Designation','family_job','Race','Sex', 'Country']

In [109]:
train_set.head()

Employment      Degree               Status         Designation  \
0          State-gov   Bachelors        Never-married        Adm-clerical   
1   Self-emp-not-inc   Bachelors   Married-civ-spouse     Exec-managerial   
2            Private     HS-grad             Divorced   Handlers-cleaners   
3            Private        11th   Married-civ-spouse   Handlers-cleaners   
4            Private   Bachelors   Married-civ-spouse      Prof-specialty   

       family_job    Race      Sex         Country  
0   Not-in-family   White     Male   United-States  
1         Husband   White     Male   United-States  
2   Not-in-family   White     Male   United-States  
3         Husband   Black     Male   United-States  
4            Wife   Black   Female            Cuba

In [110]:
for feature in train_set.columns[:]: # (every columns[:])
    print(feature,":",len(train_set[feature].unique()),'labels')
    

Employment : 9 labels
Degree : 16 labels
Status : 7 labels
Designation : 15 labels
family_job : 6 labels
Race : 5 labels
Sex : 2 labels
Country : 42 labels


In [111]:
## converting into dictionary
train_set['Country'].value_counts().to_dict()

{' United-States': 29170,
 ' Mexico': 643,
 ' ?': 583,
 ' Philippines': 198,
 ' Germany': 137,
 ' Canada': 121,
 ' Puerto-Rico': 114,
 ' El-Salvador': 106,
 ' India': 100,
 ' Cuba': 95,
 ' England': 90,
 ' Jamaica': 81,
 ' South': 80,
 ' China': 75,
 ' Italy': 73,
 ' Dominican-Republic': 70,
 ' Vietnam': 67,
 ' Guatemala': 64,
 ' Japan': 62,
 ' Poland': 60,
 ' Columbia': 59,
 ' Taiwan': 51,
 ' Haiti': 44,
 ' Iran': 43,
 ' Portugal': 37,
 ' Nicaragua': 34,
 ' Peru': 31,
 ' France': 29,
 ' Greece': 29,
 ' Ecuador': 28,
 ' Ireland': 24,
 ' Hong': 20,
 ' Cambodia': 19,
 ' Trinadad&Tobago': 19,
 ' Laos': 18,
 ' Thailand': 18,
 ' Yugoslavia': 16,
 ' Outlying-US(Guam-USVI-etc)': 14,
 ' Honduras': 13,
 ' Hungary': 13,
 ' Scotland': 12,
 ' Holand-Netherlands': 1}

- Here Indai is replaced by 100, cuba by 95, and so does other countries with there respective numbers

In [112]:
## converting into dictionary
country_map =train_set['Country'].value_counts().to_dict()

In [115]:
train_set['Country'] = train_set['Country'].map(country_map)
train_set.head(20)

Employment         Degree                  Status  \
0           State-gov      Bachelors           Never-married   
1    Self-emp-not-inc      Bachelors      Married-civ-spouse   
2             Private        HS-grad                Divorced   
3             Private           11th      Married-civ-spouse   
4             Private      Bachelors      Married-civ-spouse   
5             Private        Masters      Married-civ-spouse   
6             Private            9th   Married-spouse-absent   
7    Self-emp-not-inc        HS-grad      Married-civ-spouse   
8             Private        Masters           Never-married   
9             Private      Bachelors      Married-civ-spouse   
10            Private   Some-college      Married-civ-spouse   
11          State-gov      Bachelors      Married-civ-spouse   
12            Private      Bachelors           Never-married   
13            Private     Assoc-acdm           Never-married   
14            Private      Assoc-voc      Married-civ-spouse   
15            Private        7th-8th      Married-civ-spouse   
16   Self-emp-not-inc        HS-grad           Never-married   
17            Private        HS-grad           Never-married   
18            Private           11th      Married-civ-spouse   
19   Self-emp-not-inc        Masters                Divorced   

           Designation      family_job                 Race      Sex  Country  
0         Adm-clerical   Not-in-family                White     Male    29170  
1      Exec-managerial         Husband                White     Male    29170  
2    Handlers-cleaners   Not-in-family                White     Male    29170  
3    Handlers-cleaners         Husband                Black     Male    29170  
4       Prof-specialty            Wife                Black   Female       95  
5      Exec-managerial            Wife                White   Female    29170  
6        Other-service   Not-in-family                Black   Female       81  
7      Exec-managerial         Husband                White     Male    29170  
8       Prof-specialty   Not-in-family                White   Female    29170  
9      Exec-managerial         Husband                White     Male    29170  
10     Exec-managerial         Husband                Black     Male    29170  
11      Prof-specialty         Husband   Asian-Pac-Islander     Male      100  
12        Adm-clerical       Own-child                White   Female    29170  
13               Sales   Not-in-family                Black     Male    29170  
14        Craft-repair         Husband   Asian-Pac-Islander     Male      583  
15    Transport-moving         Husband   Amer-Indian-Eskimo     Male      643  
16     Farming-fishing       Own-child                White     Male    29170  
17   Machine-op-inspct       Unmarried                White     Male    29170  
18               Sales         Husband                White     Male    29170  
19     Exec-managerial       Unmarried                White   Female    29170

#### Advantages
- It is easy to use
- We are not increasing feature space
 #### Disadvantages
- It will not be able to handle or provide the same weight if the frequencies are same.
    

### Target Guided Ordinal Encoding
- Ordering the labels according to the target
- Replace the labels by the joint probabilty of being 1 or 0

In [116]:
import pandas as pd
df=pd.read_csv('titanic.csv', usecols=['Cabin','Survived'])
df.head()

Survived Cabin
0         0   NaN
1         1   C85
2         1   NaN
3         1  C123
4         0   NaN

In [117]:
df['Cabin'].fillna('Missing',inplace=True)

In [120]:
df.head(20)

Survived    Cabin
0          0  Missing
1          1      C85
2          1  Missing
3          1     C123
4          0  Missing
5          0  Missing
6          0      E46
7          0  Missing
8          1  Missing
9          1  Missing
10         1       G6
11         1     C103
12         0  Missing
13         0  Missing
14         0  Missing
15         1  Missing
16         0  Missing
17         1  Missing
18         0  Missing
19         1  Missing

In [122]:
df['Cabin'].astype(str).str[0] # extracting the first latter

0      M
1      C
2      M
3      C
4      M
      ..
886    M
887    B
888    M
889    C
890    M
Name: Cabin, Length: 891, dtype: object

In [124]:
df['Cabin']=df['Cabin'].astype(str).str[0]

In [125]:
df.head()

Survived Cabin
0         0     M
1         1     C
2         1     M
3         1     C
4         0     M

In [127]:
df.Cabin.unique()

array(['M', 'C', 'E', 'G', 'D', 'A', 'B', 'F', 'T'], dtype=object)

In [128]:
df.groupby(['Cabin'])['Survived'].mean()

Cabin
A    0.466667
B    0.744681
C    0.593220
D    0.757576
E    0.750000
F    0.615385
G    0.500000
M    0.299854
T    0.000000
Name: Survived, dtype: float64

- It gives the probability of survival of people in different cabins

In [132]:
df.groupby(['Cabin'])['Survived'].mean().sort_values() # sorting the values

Cabin
T    0.000000
M    0.299854
A    0.466667
G    0.500000
C    0.593220
F    0.615385
B    0.744681
E    0.750000
D    0.757576
Name: Survived, dtype: float64

In [131]:
indexing the sorting values
df.groupby(['Cabin'])['Survived'].mean().sort_values().index

Index(['T', 'M', 'A', 'G', 'C', 'F', 'B', 'E', 'D'], dtype='object', name='Cabin')

In [133]:
ordinal_labels = df.groupby(['Cabin'])['Survived'].mean().sort_values().index
ordinal_labels

Index(['T', 'M', 'A', 'G', 'C', 'F', 'B', 'E', 'D'], dtype='object', name='Cabin')

In [136]:
enumerate(ordinal_labels,0)

In [138]:
ordinal_labels2 = {k:i for i, k in enumerate(ordinal_labels,0)} # this is dictionary comprehension
ordinal_labels2

{'T': 0, 'M': 1, 'A': 2, 'G': 3, 'C': 4, 'F': 5, 'B': 6, 'E': 7, 'D': 8}

- Based on the index value we assigned the ranking to the index
- Here based on the frequency of index D which have (D = 0.757576) values  it got the highets ranking.

- The dictionary is actually the dictionary comprehension. Here enumerate(ordinal_labels,0) will iterate thorugh each and every lebels, and the first value it assign is zero.
- When you do (ordinal_label ,0), the k basically indicates the ordinal lebals(index)['T', 'M', 'A', 'G', 'C', 'F', 'B', 'E', 'D'] and i indicates the labels that assigned values to labels(0,1,2,3,4,5,6,7,8)

In [140]:
df['Cabin_ordinal_labels']=df['Cabin'].map(ordinal_labels2)
df.head()

Survived Cabin  Cabin_ordinal_labels
0         0     M                     1
1         1     C                     4
2         1     M                     1
3         1     C                     4
4         0     M                     1

### Mean Encoding

In [149]:
df.groupby(['Cabin'])['Survived'].mean().to_dict()

{'A': 0.4666666666666667,
 'B': 0.7446808510638298,
 'C': 0.5932203389830508,
 'D': 0.7575757575757576,
 'E': 0.75,
 'F': 0.6153846153846154,
 'G': 0.5,
 'M': 0.29985443959243085,
 'T': 0.0}

In [150]:
mean_ordinal=df.groupby(['Cabin'])['Survived'].mean().to_dict()

In [152]:
df['mean_ordinal_encoding'] =df['Cabin'].map(mean_ordinal)
df.head()

Survived Cabin  Cabin_ordinal_labels  mean_ordinal_encoding
0         0     M                     1               0.299854
1         1     C                     4               0.593220
2         1     M                     1               0.299854
3         1     C                     4               0.593220
4         0     M                     1               0.299854

- Prone to overfitting

### Probabilty ratio encoding

In [198]:
df = pd.read_csv('titanic.csv', usecols=['Cabin','Survived'])

In [199]:
df.head()

Survived Cabin
0         0   NaN
1         1   C85
2         1   NaN
3         1  C123
4         0   NaN

In [200]:
### Replacing
df['Cabin'].fillna('Missing',inplace=True)

In [201]:
df.head(20)

Survived    Cabin
0          0  Missing
1          1      C85
2          1  Missing
3          1     C123
4          0  Missing
5          0  Missing
6          0      E46
7          0  Missing
8          1  Missing
9          1  Missing
10         1       G6
11         1     C103
12         0  Missing
13         0  Missing
14         0  Missing
15         1  Missing
16         0  Missing
17         1  Missing
18         0  Missing
19         1  Missing

In [202]:
df['Cabin'].value_counts()

Cabin
Missing        687
C23 C25 C27      4
G6               4
B96 B98          4
C22 C26          3
              ... 
E34              1
C7               1
C54              1
E36              1
C148             1
Name: count, Length: 148, dtype: int64

In [203]:
df['Cabin'].unique() # it gives unique values of cabins

array(['Missing', 'C85', 'C123', 'E46', 'G6', 'C103', 'D56', 'A6',
       'C23 C25 C27', 'B78', 'D33', 'B30', 'C52', 'B28', 'C83', 'F33',
       'F G73', 'E31', 'A5', 'D10 D12', 'D26', 'C110', 'B58 B60', 'E101',
       'F E69', 'D47', 'B86', 'F2', 'C2', 'E33', 'B19', 'A7', 'C49', 'F4',
       'A32', 'B4', 'B80', 'A31', 'D36', 'D15', 'C93', 'C78', 'D35',
       'C87', 'B77', 'E67', 'B94', 'C125', 'C99', 'C118', 'D7', 'A19',
       'B49', 'D', 'C22 C26', 'C106', 'C65', 'E36', 'C54',
       'B57 B59 B63 B66', 'C7', 'E34', 'C32', 'B18', 'C124', 'C91', 'E40',
       'T', 'C128', 'D37', 'B35', 'E50', 'C82', 'B96 B98', 'E10', 'E44',
       'A34', 'C104', 'C111', 'C92', 'E38', 'D21', 'E12', 'E63', 'A14',
       'B37', 'C30', 'D20', 'B79', 'E25', 'D46', 'B73', 'C95', 'B38',
       'B39', 'B22', 'C86', 'C70', 'A16', 'C101', 'C68', 'A10', 'E68',
       'B41', 'A20', 'D19', 'D50', 'D9', 'A23', 'B50', 'A26', 'D48',
       'E58', 'C126', 'B71', 'B51 B53 B55', 'D49', 'B5', 'B20', 'F G63',
       'C62

In [216]:
df['Cabin']=df['Cabin'].astype(str).str[0]

In [217]:
df.head()

Survived Cabin
0         0     M
1         1     C
2         1     M
3         1     C
4         0     M

In [218]:
df['Cabin'].unique() # This is the extraction of first letters from the  unique categories of cabin

array(['M', 'C', 'E', 'G', 'D', 'A', 'B', 'F', 'T'], dtype=object)

In [220]:
df.groupby(['Cabin'])['Survived'].mean() 

Cabin
A    0.466667
B    0.744681
C    0.593220
D    0.757576
E    0.750000
F    0.615385
G    0.500000
M    0.299854
T    0.000000
Name: Survived, dtype: float64

In [221]:
prob_df = df.groupby(['Cabin'])['Survived'].mean() 

In [222]:
prob_df = pd.DataFrame(prob_df)
prob_df

Survived
Cabin          
A      0.466667
B      0.744681
C      0.593220
D      0.757576
E      0.750000
F      0.615385
G      0.500000
M      0.299854
T      0.000000

In [225]:
## probability of people who died are
prob_df['died'] = 1 - prob_df['Survived']

In [226]:
prob_df.head()

Survived      died
Cabin                    
A      0.466667  0.533333
B      0.744681  0.255319
C      0.593220  0.406780
D      0.757576  0.242424
E      0.750000  0.250000

In [229]:
# Calculating the probability ratio
prob_df['Probability_ratio'] = prob_df['Survived']/prob_df['died']
prob_df.head()

Survived      died  Probability_ratio
Cabin                                       
A      0.466667  0.533333           0.875000
B      0.744681  0.255319           2.916667
C      0.593220  0.406780           1.458333
D      0.757576  0.242424           3.125000
E      0.750000  0.250000           3.000000

In [230]:
# Converting the probability ratio into dictionary

prob_df['Probability_ratio'].to_dict()



{'A': 0.875,
 'B': 2.916666666666666,
 'C': 1.4583333333333333,
 'D': 3.125,
 'E': 3.0,
 'F': 1.6000000000000003,
 'G': 1.0,
 'M': 0.42827442827442824,
 'T': 0.0}

In [231]:
 probability_encoded= prob_df['Probability_ratio'].to_dict()

In [233]:
df['Cabin_encoded'] = df['Cabin'].map(probability_encoded) ### assigning new category
df.head()

Survived Cabin  Cabin_encoded
0         0     M       0.428274
1         1     C       1.458333
2         1     M       0.428274
3         1     C       1.458333
4         0     M       0.428274

In [234]:
df.head(20)

Survived Cabin  Cabin_encoded
0          0     M       0.428274
1          1     C       1.458333
2          1     M       0.428274
3          1     C       1.458333
4          0     M       0.428274
5          0     M       0.428274
6          0     E       3.000000
7          0     M       0.428274
8          1     M       0.428274
9          1     M       0.428274
10         1     G       1.000000
11         1     C       1.458333
12         0     M       0.428274
13         0     M       0.428274
14         0     M       0.428274
15         1     M       0.428274
16         0     M       0.428274
17         1     M       0.428274
18         0     M       0.428274
19         1     M       0.428274

##### steps
- Probability of survived based in cabin categorical feature
- Probability of not survived  (1- probability of survived)
- Probability of survived/probability of dead
- Dictionary to map cabin with probability 
- Replace with the categorical feature